### PR TITLE
bpo-31581: Reduce the number of imports for functools

### PR DESCRIPTION
**Pull request by methane**
Since functools and collections are very commonly used, it's worth enough to optimize import.

https://bugs.python.org/issue31581

### CODE REVIEW

**Comment by methane** [comment]
I added 'skip news' tag, since #1269 don't have news too.

**Review by ncoghlan — DISMISSED** [review]
The `functools.singledispatch` change seems reasonable to me, since the cost of checking the import cache twice will be cheap compared to the overhead of defining a new function to apply `@singledispatch` to.

As per the comments on the argparse PR, this should be accompanied by a comment explaining the trade-off (i.e. there are many programs that use `functools` without `singledispatch`, so we trade-off making singledispatch marginally slower for the benefit of making start-up of such applications slightly faster)

However, I'm less sure of the appropriateness of the start-up vs runtime cost trade-off for `Counter.most_common`, so I'd prefer to see that change in a separate PR.

File: Lib/functools.py [hunk 1]
```diff
@@ -753,8 +751,13 @@ def singledispatch(func):
     function acts as the default implementation, and additional
     implementations can be registered using the register() attribute of the
     generic function.
-
     """
+    # There are many programs that use functools without singledispatch, so we
+    # trade-off making singledispatch marginally slower for the benefit of
+    # making start-up of such applications slightly faster.
+    from types import MappingProxyType
```
Comment (serhiy-storchaka) [inline Lib/functools.py]: It is slightly faster to import `types` and use `types.MappingProxyType` once than import `MappingProxyType` from `types`. The `from ... import` case runs a Python code from `_bootstrap.py`, while the former case runs just a C code if the types module already is imported.

File: Lib/test/test_functools.py [hunk 2]
```diff
@@ -2033,9 +2035,9 @@ def __setitem__(self, key, value):
                 self.data[key] = value
             def clear(self):
                 self.data.clear()
-        _orig_wkd = functools.WeakKeyDictionary
+        _orig_wkd = weakref.WeakKeyDictionary
```
Comment (serhiy-storchaka) [inline Lib/test/test_functools.py]: Use `try ... finally` or `test.support.swap_attr()` for guaranteeing that weakref.WeakKeyDictionary is restored if the following checks fail.

**Comment by methane** [comment]
> However, I'm less sure of the appropriateness of the start-up vs runtime cost trade-off for Counter.most_common, so I'd prefer to see that change in a separate PR.

fixed.

**Pull request merged**